### DISCUSSION TITLE
crashloopbackoff while doing horizontal pod autoscaler HPA

### DISCUSSION THREAD

**Discussion by vishk2203** [strimzi/strimzi-kafka-operator#5448] [Q&A] [answered]
My pods are showing the status as "CrashLoopBackOff" after deploying HPA. Is there solution or precautionary action could be taken to resolve this?

![image](https://user-images.githubusercontent.com/89070106/129838160-a0206259-d2f3-4228-8613-e1ddf21000dd.png)

**Comment by scholzj** [comment, marked as answer]
HPA rarely makes sense with most Kafka brokers. And it IMHO never makes sense with ZooKeeper. If you want to scale Kafka brokers or ZooKeepers, you should do it by changing the `replicas` fields in the `Kafka` custom resource. Not by scaling the StatefulSet. That will not work short-term and the operator will anyway revert it in the next reconciliation.

**Reply by vishk2203** [reply]
Thanks for your update @scholzj . Is there any alternate option to autoscale the kafka brokers and zookeepers?

**Reply by scholzj** [reply]
Well, you should probably start by thinking about why do you actually want to autoscale them - what is the purpose, what should be the metric, what do you expect to get out of it etc.

**Reply by vishk2203** [reply]
The purpose is to autoscale the kafka brokers and zookeeper when it hits the memory utilization point. I would to like increase the pods, if it runs out of memory.

**Reply by scholzj** [reply]
But why do you think that adding more pods helps you with memory utilization?

With ZooKeeper, that will for sure not happen. Adding more nodes improves the availability. But it does not split the workload between the nodes. So I do not see why would memory consumption change significantly. From some number of noes, adding more nodes would actually hurt the performance because there will be lot more synchronization required

For Kafka, the default outcome would be again that you will have the old node unchanged and add some more free nodes. You would need to move the partition-replicas between them (which in many cases is lot of data which takes a lot of time to move). And even if you do it, Kafka uses most memory for disk cache, so over time the disk cache will fill in the memory again. 

So I think this is a wrong metric and motivation for how the software is designed.

**Reply by vishk2203** [reply]
I'm working on a data ingestion so trying to automate all the features related to it. This features mainly focuses on avoiding manual intervention.

**Reply by tomncooper** [reply]
That is a nice idea but, as Jakub said, Kafka is not some stateless service. Adding new brokers will not relieve the load on the old ones without further manual intervention to rebalance the partitions between the old and new brokers.

**Reply by tomncooper** [reply]
We have items on the road map around using our Cruise Control support to automate this balancing for adding/removing brokers but that functionality is not there yet.

**Reply by vishk2203** [reply]
Oh great. What could be the time frame for this functionality?
